core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```
@@ -1637,6 +1640,83 @@ public void sync(Translog.Location location) {
+        if (promised || promiseSync.tryAcquire()) {
+            // since we made the promise to sync we gotta do it here at least once
+            drainAndSync(candidates);
+            promiseSync.release(); // now to ensure we are passing it on we release the promise so another thread can take over
+            while (fsyncQueue.isEmpty() == false && promiseSync.availablePermits() > 0) {
+                // yet if the queue is not empty AND nobody else has yet made the promise to take over we continue syncing to disk
+                drainAndSync(candidates);
+            }
+        }
+    }
+
+    private void drainAndSync(List<Tuple<Translog.Location, Consumer<Exception>>> candidates) {
+        candidates.clear();
+        fsyncQueue.drainTo(candidates);
+        syncList(candidates);
+    }
+
+    private void syncList(List<Tuple<Translog.Location, Consumer<Exception>>> candidates) {
+        Exception exception = null;
+        if (candidates.isEmpty() == false) {
+            try {
+                final Engine engine = getEngine();
+                engine.getTranslog().ensureSynced(candidates.stream().map(Tuple::v1));
+            } catch (EngineClosedException ex) {
+                // that's fine since we already synced everything on engine close - this also is conform with the methods
+                // documentation
+            } catch (IOException ex) { // if this fails we are in deep shit - fail the request
+                logger.debug("failed to sync translog", ex);
+                // this exception is passed to all listeners - we don't retry. if this doesn't work we are in deep shit
+                exception = ex;
+            }
+        }
+        for (Tuple<Translog.Location, Consumer<Exception>> tuple : candidates) {
+            Consumer<Exception> consumer = tuple.v2();
+            try {
+                consumer.accept(exception);
+            } catch (Exception ex) {
+                logger.warn("translog sync consumer failed", ex);
```

we can just throw an assertion error here after we notified all consumers